util/topsql/reporter/single_target.go

```diff
@@ -180,111 +175,105 @@ func (ds *SingleTargetDataSink) doSend(ctx context.Context, addr string, data Re
-	return nil
+	return
 }
 
 // sendBatchSQLMeta sends a batch of SQL metas by stream.
-func (ds *SingleTargetDataSink) sendBatchSQLMeta(ctx context.Context, sqlMap *sync.Map) error {
+func (ds *SingleTargetDataSink) sendBatchSQLMeta(ctx context.Context, sqlMetas []*tipb.SQLMeta) (err error) {
+	if len(sqlMetas) == 0 {
+		return
+	}
+
 	start := time.Now()
+	sentCount := 0
+	defer func() {
+		topSQLReportSQLCountHistogram.Observe(float64(sentCount))
+
+		if err != nil {
+			reportSQLDurationFailedHistogram.Observe(time.Since(start).Seconds())
+		}
+		reportSQLDurationSuccHistogram.Observe(time.Since(start).Seconds())
+	}()
+
 	client := tipb.NewTopSQLAgentClient(ds.conn)
 	stream, err := client.ReportSQLMeta(ctx)
 	if err != nil {
 		return err
 	}
-	cnt := 0
-	sqlMap.Range(func(key, value interface{}) bool {
-		cnt++
-		meta := value.(SQLMeta)
-		sqlMeta := &tipb.SQLMeta{
-			SqlDigest:     []byte(key.(string)),
-			NormalizedSql: meta.normalizedSQL,
-			IsInternalSql: meta.isInternal,
-		}
-		if err = stream.Send(sqlMeta); err != nil {
-			return false
+
+	for _, meta := range sqlMetas {
+		if err = stream.Send(meta); err != nil {
+			return
 		}
-		return true
-	})
-	// stream.Send return error
-	if err != nil {
-		return err
+		sentCount += 1
 	}
-	topSQLReportSQLCountHistogram.Observe(float64(cnt))
+
+	// See https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream for how to avoid leaking the stream
 	_, err = stream.CloseAndRecv()
-	if err != nil {
-		reportSQLDurationFailedHistogram.Observe(time.Since(start).Seconds())
-		return err
-	}
-	reportSQLDurationSuccHistogram.Observe(time.Since(start).Seconds())
-	return nil
+	return
 }
 
 // sendBatchPlanMeta sends a batch of SQL metas by stream.
-func (ds *SingleTargetDataSink) sendBatchPlanMeta(ctx context.Context, planMap *sync.Map) error {
+func (ds *SingleTargetDataSink) sendBatchPlanMeta(ctx context.Context, planMetas []*tipb.PlanMeta) (err error) {
+	if len(planMetas) == 0 {
+		return nil
+	}
+
 	start := time.Now()
+	sentCount := 0
+	defer func() {
+		topSQLReportPlanCountHistogram.Observe(float64(sentCount))
+		if err != nil {
+			reportPlanDurationFailedHistogram.Observe(time.Since(start).Seconds())
+		}
+		reportPlanDurationSuccHistogram.Observe(time.Since(start).Seconds())
+	}()
+
 	client := tipb.NewTopSQLAgentClient(ds.conn)
 	stream, err := client.ReportPlanMeta(ctx)
 	if err != nil {
 		return err
 	}
-	cnt := 0
-	planMap.Range(func(key, value interface{}) bool {
-		planDecoded, errDecode := ds.decodePlan(value.(string))
-		if errDecode != nil {
-			logutil.BgLogger().Warn("[top-sql] decode plan failed", zap.Error(errDecode))
-			return true
-		}
-		cnt++
-		planMeta := &tipb.PlanMeta{
-			PlanDigest:     []byte(key.(string)),
-			NormalizedPlan: planDecoded,
-		}
-		if err = stream.Send(planMeta); err != nil {
-			return false
+
+	for _, meta := range planMetas {
+		if err = stream.Send(meta); err != nil {
```

Comment:
Do we need some kind of batch mechanism or flow control? As now there can be a lot. @crazycs520

Reply:
https://github.com/pingcap/tidb/issues/30787

Opened an issue to track that. Without any tracking seems that the work can be easily forgotten.